backend/lib/services/shoots.js

```diff
@@ -331,9 +331,35 @@ async function getSecret (client, { namespace, name }) {
   }
 }
 
+async function getSecretByLabelSelector (client, namespace, labelSelector) {
+  try {
+    const query = {
+      labelSelector: labelSelector.join(',')
+    }
+    const secretList = await client.core.secrets.list(namespace, query)
+    return _.head(_.orderBy(secretList.items, ['metadata.creationTimestamp'], ['desc']))
+  } catch (err) {
+    logger.error('failed to fetch %s secret: %s', name, err)
+    throw err
+  }
+}
+
 async function assignMonitoringSecret (client, data, namespace, shootName) {
-  const name = shootName ? `${shootName}.monitoring` : 'monitoring-ingress-credentials'
-  const secret = await getSecret(client, { namespace, name })
+  let secret
+  if (shootName) {
```

Comment:
I tried to refactor it but this make the code even more complicated in my point of view. The code you suggested is lacking the fallback. Adding duplicates error handling without a dedicated getSecret merthod. I suggest let's keep it and refactor the whole assignMonitoring secret stuff when we remove the fallback, which we can to with one of the next releases.

Reply:
I don't agree. The fallback is of course in my code. The error handling is not duplicated. It has the error handling of `getSecret` and `getSecretByLabelSelector` in one function. It is not working because it should be `.chain(secretList.items)` and not `.chain(secretList)`. If you think we should keep the `getSecret` function we can do it like this:

```js
async function getMonitoringSecret (client, { namespace, name }) {
  if (!name.endsWith('.monitoring')) {
    try {
      const labelSelector = 'name=observability-ingress,managed-by=secrets-manager,manager-identity=gardenlet'
      const secretList = await client.core.secrets.list(namespace, { labelSelector })
      const secret = _
        .chain(secretList.items)
        .orderBy(['metadata.creationTimestamp'], ['desc'])
        .head()
        .value()
      if (secret) {
        return secret
      }
    } catch (err) {
      logger.error('failed to fetch %s secret: %s', name, err)
      throw err
    }
  }
  return getSecret(client, { namespace, name })
}
```

Reply:
yes, you are right. we can change it as you suggested. But I still think it is better to have the fallback more explicit at least we should add some comments